### PR TITLE
perf: improve how delegations and rewards are updated

### PR DESCRIPTION
**Pull request by rmontagnin**
## Description
This PR implements a better way of updating delegation rewards and delegations themselves. Instead of relying on multiple queries (delete + insert), we replaced everything with a single [transaction](https://www.postgresql.org/docs/current/tutorial-transactions.html). This should improve the performance of the updates as well as provide a fix for empty events being sent to Hasura when such updates are performed (ref forbole/forbole-x#503). Using a single transaction should update all the rows at the same time, without any empty event in between. 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch
- [ ] provided a link to the relevant issue or specification
- [x] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: database/staking_delegations.go [hunk 1]
```diff
@@ -116,31 +143,70 @@ func (db *Db) GetUserDelegationsAmount(address string) (sdk.Coins, error) {
...
+	err = db.saveDelegationsWithTx(tx, height, delegations)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	// Commit the transaction
+	err = tx.Commit()
+	if err != nil {
+		return fmt.Errorf("error while committing validator deelgations transactionL %s", err)
```
Comment (huichiaotsou) [inline database/staking_delegations.go]:
```suggestion
		return fmt.Errorf("error while committing validator deelgations transaction %s", err)
```

File: database/staking_delegations.go [hunk 2]
```diff
@@ -116,31 +143,70 @@ func (db *Db) GetUserDelegationsAmount(address string) (sdk.Coins, error) {
...
 DELETE FROM delegation USING validator_info 
 WHERE delegation.validator_address = validator_info.consensus_address 
-  AND validator_info.operator_address = $1`
-
-	_, err := db.Sql.Exec(stmt, valOperAddr)
+  AND validator_info.operator_address = $1 AND delegation.height <= $2`
+	_, err = tx.Exec(stmt, valOperAddr, height)
 	if err != nil {
+		tx.Rollback()
 		return fmt.Errorf("error while deleting delegations for valdiator: %s", err)
```
Comment (MonikaCat) [inline database/staking_delegations.go]:
```suggestion
		return fmt.Errorf("error while deleting delegations for validator: %s", err)
```

**Review by huichiaotsou — COMMENTED** [review]
While I was testing locally, the BDjuno would be stuck at some point. I guess there could be a deadlock issue but I haven't figured out where..

**Comment by huichiaotsou** [comment]
> While I was testing locally, the BDjuno would be stuck at some point. I guess there could be a deadlock issue but I haven't figured out where..

Hmmm I guess it's the go routines at modules/distribution/utils_delegator_rewards.go

**Review by MonikaCat — APPROVED** [review]
Tested, LGTM

**Comment by MonikaCat** [comment]
@huichiaotsou Oh really? Are you testing it using local or remote node? Is there any error being thrown?

**Comment by rmontagnin** [comment]
@huichiaotsou What chain are you trying to parse?

**Comment by huichiaotsou** [comment]
@MonikaCat  my local node. no error, bdjuno just stopped there 
@RiccardoM desmos v1, from block 1

**Comment by huichiaotsou** [comment]
> @MonikaCat my local node. no error, bdjuno just stopped there @RiccardoM desmos v1, from block 1

yeah after removing go routines from [here](https://github.com/forbole/bdjuno/blob/aabff5dd006741e0268d540610ee77db9e8483ee/modules/distribution/utils_delegator_rewards.go#L32) it no longer stops

**Comment by rmontagnin** [comment]
Thanks @huichiaotsou for identifying the issue. I've now deleted the concurrent delegator rewards update to solve this issue. I personally don't think this will cause any major performance downgrade due to the fact that querying + storing a delegator reward is very fat anyway.

**Review by huichiaotsou — APPROVED** [review]

**Pull request merged**